File: CodeWars_Ruby_solution/5kyu/Directions Reduction.rb
def dirReduc(arr)
    i = 0
    while (i < arr.length - 1)
      case arr[i]
      when "EAST"
        if arr[i + 1] == "WEST"
          arr.delete_at(i + 1)
          arr.delete_at(i)
          i -= 1 if i > 0
          next
        end
      when "SOUTH"
        if arr[i + 1] == "NORTH"
          arr.delete_at(i + 1)
          arr.delete_at(i)
          i -= 1 if i > 0
          next
        end
      when "WEST"
        if arr[i + 1] == "EAST"
          arr.delete_at(i + 1)
          arr.delete_at(i)
          i -= 1 if i > 0
          next
        end
      when "NORTH"
        if arr[i + 1] == "SOUTH"
          arr.delete_at(i + 1)
          arr.delete_at(i)
          i -= 1 if i > 0
          next
        end
      end
      i += 1
    end
    arr
  end

File: CodeWars_Cpp_solution/5 kyu/Directions Reduction.cpp
class DirReduction
{
public:
    static std::vector<std::string> dirReduc(std::vector<std::string> &arr){
      int i, len;
      for (i = 0, len = arr.size() - 1; i < len; ){
        if (arr[i] == "EAST"){
          if (arr[i + 1] == "WEST"){
            arr.erase(arr.begin() + i, arr.begin() + i + 2);
            len -= 2;
            if (i > 0)
              i--;
            continue;
          }
        }
        else if (arr[i] == "SOUTH"){
          if (arr[i + 1] == "NORTH"){
            arr.erase(arr.begin() + i, arr.begin() + i + 2);
            len -= 2;
            if (i > 0)
              i--;
            continue;
          }
        }
        else if (arr[i] == "WEST"){
          if (arr[i + 1] == "EAST"){
            arr.erase(arr.begin() + i, arr.begin() + i + 2);
            len -= 2;
            if (i > 0)
              i--;
            continue;
          }
        }
        else if (arr[i] == "NORTH"){
          if (arr[i + 1] == "SOUTH"){
            arr.erase(arr.begin() + i, arr.begin() + i + 2);
            len -= 2;
            if (i > 0)
              i--;
            continue;
          }
        }
        i++;
      }
      return arr;
    }
};
